File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        
        int n = s.length();
        int left = 0;
        int[] hash = new int[26];
        int maxf = 0;
        int maxLen = 0;
        for(int right=0;right<n;right++){
            hash[s.charAt(right) - 'A']++;

            maxf = Math.max(maxf, hash[s.charAt(right) - 'A']);

            int changes = (right - left +1) - maxf;

            if(changes > k){
                hash[s.charAt(left) - 'A']--;
                left++;
            }
            maxLen = Math.max(maxLen, right -left +1);
        }

        return maxLen;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {

    let freq = new Array(26).fill(0); 
    let left = 0;
    let maxFreq = 0;
    let maxWindow  = 0;

    for(let right = 0;right < s.length;right++){
        freq[s.charCodeAt(right) - 65]++;

        maxFreq = Math.max(maxFreq, freq[s.charCodeAt(right) - 65]);

        let windowLength = right - left + 1;

        if(windowLength - maxFreq > k){
            freq[s.charCodeAt(left) - 65]--; 
            left++;
        }

        windowLength = right - left +1;
        maxWindow = Math.max(maxWindow, windowLength);
    }

    return maxWindow;
    
};